### DISCUSSION TITLE
Executing multiple statements in `session.execute`

### DISCUSSION THREAD

**Discussion by faulter-cronkite** [sqlalchemy/sqlalchemy#8193] [Usage Questions] [answered]
Concurrent `INSERT INTO`s are not safe for parts of my application, so I'm using PG's lovely txn advisory locks, i.e.:

```python
import sqlalchemy
import sqlalchemy.dialects.postgresql as postgresql

lock_fn = sqlalchemy.func.pg_advisory_xact_lock(1234)

# acquire lock
session.execute(sqlalchemy.select(lock_fn))

# insert into some table
insert = postgresql.insert(SomeTable)
insert = insert.values(some_list_of_dicts)

session.execute(insert)
```

This works wonderfully, but I'd really like to avoid the lock round trip. At the same time, I'd like to avoid constructing the `INSERT`s by hand, because the system has many tables and the `postgresql.insert` factory is super handy.

Ideally I'd call `session.execute` once, with two statements: the lock, then the insert. I confirm that something like `session.execute("SELECT pg_advisory_xact_lock(1234); INSERT INTO some_table (field) VALUES ('something')")` works, but constructing this string has been... difficult. I've essentially tried this:

```python
compiled = insert.compile(
    dialect=postgresql.dialect(
        paramstyle='named'
    )
)

statement = "SELECT pg_advisory_xact_lock(1234); " + compiled.string

session.execute(statement, compiled.params)
```

But now the compiled statement has some oddities of `paramstyle` mismatches:

```
INSERT INTO some_table(field_a, field_b)
VALUES
(%(field_a_m0), :field_b_m0::text[]),
(%(field_a_m0), :field_b_m0::text[])
;
```

amongst other issues with the compiled statement (i.e. type mismatches).

I would love to be able to bundle statements together in a single `session.execute` call. Has anyone done anything similar? Any tips would be hugely appreciated!

**Comment by zzzeek** [comment, marked as answer]
there's two separate answers to this.

first is, I really think you should have a proven rationale for your request to merge the two statements together, by which I mean you would measure the performance advantage you get from this.     not all drivers support the semicolon thing in any case.     The argument that could be made for that is likely that SQLAlchemy's own statement execution mechanics are expensive here; ``cursor.execute()`` on the DBAPI connection will not be expensive at all, and there are ways to get at that cursor.     However, it doesn't matter much because the solution to "use a semicolon" vs. "execute on the cursor" look 95% the same anyway, see next answer

second, as this is a structural thing you want to add to statements (or as a prior execution), this should definitely be a [before_cursor_execute](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=before_cursor_execute#sqlalchemy.events.ConnectionEvents.before_cursor_execute) thing.

it's easier to type out so here are both approaches, note we use execution_options to invoke the command, that's what execution_options is for.

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import insert
from sqlalchemy import Integer
from sqlalchemy import event
from sqlalchemy import String
import sqlalchemy.dialects.postgresql as postgresql
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)


e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)


@event.listens_for(e, "before_cursor_execute", retval=True)
def _set_advisory_lock(
    conn, cursor, statement, parameters, context, executemany
):
    advisory_lock = context.execution_options.get("pg_advisory_lock", None)
    if advisory_lock:

        # LIKELY PREFERABLE, BETTER FOR THE DRIVER, SHOULD PERFORM VERY WELL:
        # invoke on the DBAPI cursor.  this already has an open channel
        # to the DB so the overhead is a handful of C functions for
        # psycopg2
        # cursor.execute("SELECT pg_advisory_xact_lock(%s)", (advisory_lock,))

        # CONCATENATE INSTEAD.
        # BE CAREFUL of SQL injection, ensure input is trusted
        statement = (
            f"SELECT pg_advisory_xact_lock({advisory_lock}); {statement}"
        )

        return statement, parameters


session = Session(e)
# insert into some table
insert = postgresql.insert(A)
insert = insert.values([{"id": 1, "data": "one"}, {"id": 2, "data": "two"}])

session.execute(insert, execution_options={"pg_advisory_lock": 1234})
```

**Reply by faulter-cronkite** [reply]
Wow, beautiful, thank you so much @zzzeek . The `event` API comes to the rescue again. I'm marking this as the answer, and will be taking your advice and invoking the cursor.

Thank you for all of your great work on this library!
